### PR TITLE
Added edge cost routine for total cost in edit_distance()

### PR DESCRIPTION
**Pull request by ChamanAgrawal**
Fix: #1036

### CODE REVIEW

**Comment by juliohm** [comment]
Hi @ChamanAgrawal , I am not sure of the changes you are proposing. They seem too radical at first?

Can you please try to refactor the code to facilitate our review? Also, the fact that the function `edge_cost` is expecting 6 arguments is a good indication that the code should be organized differently.

Please consider sharing some of your thoughts before refactoring the code. That will help us understand what you are trying to do, and will help us in the review later.

**Comment by ChamanAgrawal** [comment]
Overview of the implementation:
We will edit an edge only when both of its vertices are edited so when a vertex u is edited we search in its neighbors for already edited vertices, for every such vertex v we edit the edge (u, v).

Note: f: u->v represents u in G1 is edited (mapped) to v in G2.

Classifying the edge edits:
1) Edge deletion: An edge is considered deleted if (u, v) is in G1 but (f(u), f(v)) is not in G2. So in ` edit_cost() ` if either the current vertex ` λ⁺[end][2] `is deleted or the neighbour ` v ` is already deleted or vertices are not deleted in edit but edge is deleted ` !(has_edge(G₂, v, λ⁺[end][2]) || has_edge(G₂, λ⁺[end][2], v) ` then ` edge_delete_cost() ` will give the cost.
2) If in the above search in neighbors of ` u ` an edge is not deleted that means it is substituted so ` edge_subst_cost() ` will give the cost.
3) It is also possible that new edges may have been added to f(u) i.e. ` λ⁺[end][2] ` so we will search in the neighbor of ` λ⁺[end][2] ` for already edited vertices v in G2 and check for corresponding vertices f^-1(v) in G1 to be neighbor of u , if they are not i.e. ` !(has_edge(G₁, u, λ⁺[end][1]) || has_edge(G₁, λ⁺[end][1], u)) ` then edge has been added so ` edge_insert_cost() ` will give cost.

Now for the changes in the ` # add remaining vertices of G₂ to the path ` part.
Let's call the already edited vertices set1 and the remaining vertices to be added as set2. Unlike the case when edges were ignored this time we can not add all the remaining vertices directly as vertices in set2 will have edges among them self and with set1 to be considered. So to be consistent with our defined procedure for edge edit ( "We will edit an edge only when both of its vertices are edited so when a vertex u is edited we search in its neighbors for already edited vertices, for every such vertex v we edit the edge (u, v)." )  while adding each vertex we will search for already edited vertices in its neighbors in G2 and add edge accordingly as here.
```
                for v in vs
                    total_insert_cost += insert_cost(v)
                    u_neighbors = is_directed(G₂) ? (inneighbors(G₂, v)..., outneighbors(G₂, v)...) : neighbors(G₂, v)
                    for u in u_neighbors
                        for (w, x) in λ⁺
                            if x == u
                                total_insert_cost += edge_insert_cost(w, u)
                            end
                        end
                    end
                    λ⁺ = [λ⁺; (0, v)]
                end
```

**Comment by sbromberger** [comment]
@juliohm - I'm setting you as reviewer. Thanks.

**Comment by juliohm** [comment]
Thank you @sbromberger, I should really reserve some time for it. :+1: I will set a deadline in my calendar for January. I need to review the algorithm as described in the book, and see if I can work together with @ChamanAgrawal to make the code look clearer.

Happy New Year to all of you! :fireworks: 

Greeting from Rio,

-Julio

**Comment by juliohm** [comment]
Oh man, I am trying to find the time to review this, but still couldn't find it! Please apologize my delay, I am sure I will be able to look at it at some point. :pray:

**Comment by juliohm** [comment]
@ChamanAgrawal I finally had a chance to look into the PR, and based on your comments, I think that we can refactor the code to make it cleaner and more efficient.

It seems that the edge costs can be computed as a post-processing step, but please correct me if I am wrong. Instead of modifying the algorithm, we could compute a path + cost with the existing implementation, and then post-process the resulting path once to find out which edges were modified. Makes sense?

**Comment by juliohm** [comment]
Actually, the resulting path may change if we include the edges costs right? I think my previous comment is incorrect.

File: src/edit_distance.jl [hunk 1]
```diff
@@ -80,20 +83,76 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
 end
 
+function edge_cost(G₁::AbstractGraph, G₂::AbstractGraph, λ⁺,
+                    edge_insert_cost::Function,
+                    edge_delete_cost::Function,
+                    edge_subst_cost::Function)
+
+    cu = []
+    cv = []
+    u_neighbors = is_directed(G₁) ? (inneighbors(G₁, λ⁺[end][1])..., outneighbors(G₁, λ⁺[end][1])...) : neighbors(G₁, λ⁺[end][1])
```
Comment (juliohm) [inline src/edit_distance.jl]: Use `all_neighbors`

File: src/edit_distance.jl [hunk 2]
```diff
@@ -80,20 +83,76 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
+    if λ⁺[end][2]!=0
+        v_neighbors = is_directed(G₂) ? (inneighbors(G₂, λ⁺[end][2])..., outneighbors(G₂, λ⁺[end][2])...) : neighbors(G₂, λ⁺[end][2])
```
Comment (juliohm) [inline src/edit_distance.jl]: Use `all_neighbors`

File: src/edit_distance.jl [hunk 3]
```diff
@@ -80,20 +83,76 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
         end
     end
 end
 
+function edge_cost(G₁::AbstractGraph, G₂::AbstractGraph, λ⁺,
+                    edge_insert_cost::Function,
+                    edge_delete_cost::Function,
+                    edge_subst_cost::Function)
+
+    cu = []
```
Comment (juliohm) [inline src/edit_distance.jl]: If this is a vector of `Float64`, help the compiler by providing the type:

```
cu = Vector{Float64}()
```

File: src/edit_distance.jl [hunk 4]
```diff
@@ -80,20 +83,76 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
+    for v in v_neighbors
+        for (u, w) in λ⁺[1:end-1]
+            if w == v
+                cv = [cv; (u, w)]
+            end
+        end
+    end
+    edgecost = 0
+    for (u,v) in cu
+        if (v == 0 || λ⁺[end][2] == 0) || !(has_edge(G₂, v, λ⁺[end][2]) || has_edge(G₂, λ⁺[end][2], v))
```
Comment (juliohm) [inline src/edit_distance.jl]: This is not easy to read. Could you maybe refactor the code and give names to some of these variables? `lambda[end][2]` for example is not good.

**Comment by ChamanAgrawal** [comment]
@juliohm Yes, the resulting path will change if we include the edges costs. I will do the required changes and add some comments within 2 days. My mid-semester exams start next week so I am pretty busy but I will try.

File: src/edit_distance.jl (hunk 3 above)
Comment (ChamanAgrawal) [reply src/edit_distance.jl]: No, it contains edges.

Comment (juliohm) [reply src/edit_distance.jl]: Then, you could provide the correct type `Vector{Tuple{Int,Int}}()` It should be better than leaving it as a vector of `Any`.

**Comment by ChamanAgrawal** [comment]
Instead of using a new function for edge cost I have mixed it with the main function so no more function calls with too many arguments. Also, I have changed the messy iteration in neighbors for checking them to be transformed by using has_edge().

File: src/edit_distance.jl [hunk 5]
```diff
@@ -78,17 +81,83 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
             if k < nv(G₁) # there are still vertices to process in G₁?
+                # u be the next node to be transformed
+                u = k+1
+                # transformation operations on neighbor of u 
+                u_edited_neighbors = Vector{Tuple{Int,Int}}()
+                u_neighbors = all_neighbors(G₁, u)
+                
+                # # add transformation (x, y) to u_edited_neighbors if x is neighbor of u
+                for (x, y) in λ
+                    if has_edge(G₁, k+1, x) || has_edge(G₁, x, k+1)
```
Comment (juliohm) [inline src/edit_distance.jl]: Should `k+1` be replaced by `u` since you gave this vertex a name?

File: src/edit_distance.jl [hunk 6]
```diff
@@ -78,17 +81,83 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
             vs = setdiff(1:nv(G₂), [v for (u, v) in λ])
 
             if k < nv(G₁) # there are still vertices to process in G₁?
+                # u be the next node to be transformed
+                u = k+1
+                # transformation operations on neighbor of u 
+                u_edited_neighbors = Vector{Tuple{Int,Int}}()
+                u_neighbors = all_neighbors(G₁, u)
+                
+                # # add transformation (x, y) to u_edited_neighbors if x is neighbor of u
```
Comment (juliohm) [inline src/edit_distance.jl]: So `u_dited_neighbors` holds neighbors or transformations? It is a little confusing the variable content does not match its name.

File: src/edit_distance.jl [hunk 7]
```diff
@@ -78,17 +81,83 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
             vs = setdiff(1:nv(G₂), [v for (u, v) in λ])
 
             if k < nv(G₁) # there are still vertices to process in G₁?
+                # u be the next node to be transformed
+                u = k+1
+                # transformation operations on neighbor of u 
+                u_edited_neighbors = Vector{Tuple{Int,Int}}()
+                u_neighbors = all_neighbors(G₁, u)
```
Comment (juliohm) [inline src/edit_distance.jl]: So this variable is not used anywhere? Can you also double check `v_neighbors`? Please reserve some time to carefully review your code before asking for new review. :+1:

File: src/edit_distance.jl [hunk 8]
```diff
@@ -78,17 +81,83 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
+                
+                # # add transformation (x, y) to u_edited_neighbors if x is neighbor of u
+                for (x, y) in λ
```
Comment (juliohm) [inline src/edit_distance.jl]: I would replace this entire for loop by a simple `filter` call:

```julia
edited = filter(x -> has_edge(G₁, u, x) || has_edge(G₁, x, u), λ)
```

File: src/edit_distance.jl (hunk 6 above)
Comment (ChamanAgrawal) [reply src/edit_distance.jl]: It holds the transformation, yes the name is indeed confusing

File: src/edit_distance.jl (hunk 7 above)
Comment (ChamanAgrawal) [reply src/edit_distance.jl]: Okay, thanks for the review. I will carefully plan and review my code.

**Comment by ChamanAgrawal** [comment]
@juliohm I don't think there is repetition here except between the two loops with nearly similar conditions at line 112 and 123 but they are just loops. Actually earlier when I made a function for this, some computations were unnecessarily repeated due to multiple function calls. 
I have worked on the other comments, can you please guide me through an example of repetition in the code when you are free.

File: src/edit_distance.jl [hunk 9]
```diff
@@ -78,17 +81,83 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
+
+                    # add transformation (x, y) to v_edited_neighbors if y is neighbor of v
+                    for (x, y) in λ
+                        if has_edge(G₂, v, y) || has_edge(G₂, y, v)
+                            v_edited_neighbors = [v_edited_neighbors; (x, y)]
+                        end
+                    end
+
+                    # add edge transformation cost
+                    for (x, y) in u_edited_neighbors
```
Comment (ChamanAgrawal) [inline src/edit_distance.jl]: similar loops: first

File: src/edit_distance.jl [hunk 10]
```diff
@@ -78,17 +81,83 @@ function edit_distance(G₁::AbstractGraph, G₂::AbstractGraph;
...
+                        if (y == 0) || !(has_edge(G₂, y, v) || has_edge(G₂, v, y))
+                            edgecost += edge_delete_cost(x, u)
+                        # if node is neighbor even after transformation
+                        else
+                            edgecost += edge_subst_cost((x, u), (y, v))
+                        end
+                    end
+
+                    # new neighbors formed after transformation
+                    for (x, y) in v_edited_neighbors
```
Comment (ChamanAgrawal) [inline src/edit_distance.jl]: similar loop: second

**Comment by sbromberger** [comment]
Hi folks,

Where do we stand on this PR? Can we move forward or is it stalled?

**Comment by juliohm** [comment]
Hi Seth, as far as I remember, it is not in the right shape to merge yet.
I'm in a conference in Vienna this week with limited internet access. But I
can try take a look next Monday.

**Pull request closed without merging**